boot/systems.go

```diff
@@ -0,0 +1,133 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// clearTryRecoverySystem removes a given candidate recovery system from the
+// modeenv state file, reseals and clears related bootloader variables.
+func clearTryRecoverySystem(dev Device, systemLabel string) error {
+	if !dev.HasModeenv() {
+		return fmt.Errorf("internal error: recovery systems can only be used on UC20")
+	}
+
+	m, err := loadModeenv()
+	if err != nil {
+		return err
+	}
+	opts := &bootloader.Options{
+		// setup the recovery bootloader
+		Role: bootloader.RoleRecovery,
+	}
+	bl, err := bootloader.Find(InitramfsUbuntuSeedDir, opts)
+	if err != nil {
+		return err
+	}
+
+	found := false
+	for idx, sys := range m.CurrentRecoverySystems {
+		if sys == systemLabel {
+			found = true
+			m.CurrentRecoverySystems = append(m.CurrentRecoverySystems[:idx],
+				m.CurrentRecoverySystems[idx+1:]...)
+			break
+		}
+	}
+	if found {
+		// we may be repeating the cleanup, in which case the system may
+		// not be present in modeenv already
+		if err := m.Write(); err != nil {
+			return err
+		}
+	}
+	// but we still want to reseal, in case the cleanup did not reach this
+	// point before
+	const expectReseal = true
+	resealErr := resealKeyToModeenv(dirs.GlobalRootDir, dev.Model(), m, expectReseal)
+
+	// clear both variables, no matter the values they hold
+	vars := map[string]string{
+		"try_recovery_system":    "",
+		"recovery_system_status": "",
+	}
+	// try to clear regardless of reseal failing
+	blErr := bl.SetBootVars(vars)
```

Comment:
speaking of this kind of ordering, what is your opinion on how hard it would be to make something in `resealKeyToModeenv` `panic()` thus simulating a reboot in the middle of the operation to ensure that operations are in a consistent state? too much work / too little reward since it's a fairly simple set of operations right now?

Reply:
I think Ian is right here about the order unless we are confused

Reply:
Booting into the system is controlled by `snapd_recovery_mode` and `snapd_recovery_system`, which are set separately. This cleanup is invoked if SetTryRecoverySystem fails, which is before we try to reboot to a new system, or after we're back to run mode, in which case the bootloader variables selecting the mode/system would have been modified and would no longer boot into the tried system. Need to think about it a bit, maybe I missed something.